File: Data Structures & Algorithms/single-number/submission-1.java
class Solution {
    public int singleNumber(int[] nums) {
        // 由於每個值會出現兩次，所以只要找到哪個 bit manipulation 可以做到相同值為 0
        // 的操作，而 XOR 可以做到這樣
        // 然後由於 XOR 具有交換性的操作，所以 [3, 2, 3] 可以轉換成 [3, 3, 2]
        // 3 ^ 3 = 0, 0 ^ 2 = 2
        
        int res = nums[0];
        for (int i = 1; i < nums.length; ++i) {
            res ^= nums[i];
        }
        return res; 
    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1] * len(nums)
        # nums = [1, 2, 3, 4]
        for i in range(1, len(nums)):
            # res = [1, 1, 2, 6]
            res[i] = res[i - 1] * nums[i - 1]
        postfix = 1
        for i in range(len(nums) - 1, -1, -1):
            res[i] = res[i] * postfix
            # posfitx = [24, 12, 4, 1]
            postfix = postfix * nums[i]
            print(postfix)
        return res